streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -671,18 +688,35 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
   final def flatMap[R1 <: R, E1 >: E, B](f0: A => ZStream[R1, E1, B]): ZStream[R1, E1, B] =
     new ZStream[R1, E1, B] {
       def fold[R2 <: R1, E2 >: E1, B1 >: B, S]: Fold[R2, E2, B1, S] =
-        ZManaged.succeed { (s, cont, f) =>
-          self.fold[R2, E2, A, S].flatMap { foldOuter =>
-            foldOuter(s, cont, (s, a) => {
-              f0(a)
-                .fold[R2, E2, B1, S]
-                .flatMap { foldInner =>
-                  foldInner(s, cont, f)
+        foldDefault
+
+      override def process: ZManaged[R1, E1, ZStream.InputStream[R1, E1, B]] =
```

Comment:
@jdegoes could you review this?

Reply:
About as terrible as I suspected `flatMap` would be... 😆  not all of course will increase, on average I expect we'll come out with much less code.